jetty-websocket/websocket-core/src/test/java/org/eclipse/jetty/websocket/testimpl/WebSocketHandler.java

```diff
@@ -0,0 +1,457 @@
+import org.eclipse.jetty.util.BufferUtil;
+import org.eclipse.jetty.util.Callback;
+import org.eclipse.jetty.util.DecoratedObjectFactory;
+import org.eclipse.jetty.util.component.ContainerLifeCycle;
+import org.eclipse.jetty.util.log.Log;
+import org.eclipse.jetty.util.log.Logger;
+import org.eclipse.jetty.websocket.core.Frame;
+import org.eclipse.jetty.websocket.core.ParserDeMasker;
+import org.eclipse.jetty.websocket.core.WebSocketBehavior;
+import org.eclipse.jetty.websocket.core.WebSocketCoreSession;
+import org.eclipse.jetty.websocket.core.WebSocketLocalEndpoint;
+import org.eclipse.jetty.websocket.core.WebSocketPolicy;
+import org.eclipse.jetty.websocket.core.WebSocketRemoteEndpoint;
+import org.eclipse.jetty.websocket.core.extensions.ExtensionConfig;
+import org.eclipse.jetty.websocket.core.extensions.ExtensionStack;
+import org.eclipse.jetty.websocket.core.extensions.WebSocketExtensionRegistry;
+import org.eclipse.jetty.websocket.core.frames.TextFrame;
+import org.eclipse.jetty.websocket.core.handshake.AcceptHash;
+import org.eclipse.jetty.websocket.core.handshake.UpgradeRequest;
+import org.eclipse.jetty.websocket.core.handshake.UpgradeResponse;
+import org.eclipse.jetty.websocket.core.io.WebSocketCoreConnection;
+import org.eclipse.jetty.websocket.core.io.WebSocketRemoteEndpointImpl;
+
+import javax.servlet.ServletContext;
+import javax.servlet.ServletException;
+import javax.servlet.http.HttpServletRequest;
+import javax.servlet.http.HttpServletResponse;
+import java.io.IOException;
+import java.nio.ByteBuffer;
+import java.util.Collections;
+import java.util.List;
+import java.util.stream.Collectors;
+
+public class WebSocketHandler extends HandlerWrapper
+{
+    static final Logger LOG = Log.getLogger(WebSocketHandler.class);
+
+    private static HttpField UpgradeWebSocket = new PreEncodedHttpField(HttpHeader.UPGRADE,"WebSocket");
+    private static HttpField ConnectionUpgrade = new PreEncodedHttpField(HttpHeader.CONNECTION,HttpHeader.UPGRADE.asString());
+    private static Handshaker RFC6455_Handshaker = new RFC6455Handshaker();
+
+
+    public interface Handshaker
+    {
+        boolean upgradeRequest(Request baseRequest, HttpServletRequest request, HttpServletResponse response) throws IOException;
+    }
+
+    interface WebSocketConnectionFactory extends ConnectionFactory
+    {
+        WebSocketCoreConnection newConnection(Connector connector, EndPoint endPoint, List<String> extensions);
+
+        @Override
+        default Connection newConnection(Connector connector, EndPoint endPoint)
+        {
+            throw new UnsupportedOperationException();
+        }
+
+        @Override
+        default String getProtocol()
+        {
+            return "ws";
+        }
+
+        @Override
+        default List<String> getProtocols()
+        {
+            return Collections.singletonList(getProtocol());
+        }
+    }
+
+    interface WebSocketSessionFactory
+    {
+        // TODO why is core session abstract?
```

Comment:
WebSocketCoreSession is not exposed via the WebAppClassLoader, it is not an end user API, the specific APIs need to extend from WebSocketCoreSession to make their own APIs work successfully through the WebAppClassLoader.

Reply:
They may extend, they may wrap, they may map.  The APIs should be free to use the core session however they want.    There are no abstract methods on core and it appears functional as is, so it is not abstract.